### PR TITLE
fix(enterprise/coderd): check provisionerd API version on connection

### PR DESCRIPTION
**Pull request by johnstcn**
Extracted from https://github.com/coder/coder/pull/12021

Moves package `coderd/util/apiversion` to `apiversion` to avoid import cycles. 
Adds a version check when a provisioner daemon attempts to connect to coderd.
  We were not doing this before.
  We assume that the API version is "1.0" if no version is specified.
Adds a test that a missing `version` parameter is treated as API version 1.0.

### CODE REVIEW

**Review by mafredri — APPROVED** [review]
> Moves package coderd/util/apiversion to apiversion to avoid import cycles.

I'm not opposed to the change but I saw nothing that would've cause an import cycle without the restructuring.

File: codersdk/provisionerdaemons.go [hunk 1]
```diff
@@ -193,14 +194,17 @@ type ServeProvisionerDaemonRequest struct {
 // ServeProvisionerDaemon returns the gRPC service for a provisioner daemon
 // implementation. The context is during dial, not during the lifetime of the
 // client. Client should be closed after use.
-func (c *Client) ServeProvisionerDaemon(ctx context.Context, req ServeProvisionerDaemonRequest) (proto.DRPCProvisionerDaemonClient, error) {
+func (c *Client) ServeProvisionerDaemon(ctx context.Context, v *apiversion.APIVersion, req ServeProvisionerDaemonRequest) (proto.DRPCProvisionerDaemonClient, error) {
```
Comment (aslilac) [inline codersdk/provisionerdaemons.go:197]: I'm not crazy about this parameter. it makes it easier to do the wrong thing (not pass a version) than the right thing (specific `CurrentVersion`).

Comment (johnstcn) [reply codersdk/provisionerdaemons.go:197]: Could refactor to something like the below?

```
func (c *Client) ServeProvisionerDaemon(ctx, context.Context, req ServeProvisionerDaemonRequest)  {
   serveProvisionerDaemonWithVersion(ctx, currentVersion, req)
}
```

Comment (johnstcn) [reply codersdk/provisionerdaemons.go]: I ended up restoring the original method signature and just hitting the endpoint with a plain request, which accomplishes the same goal but in a slightly wordier fashion.

**Review by aslilac — APPROVED** [review]

**Pull request merged**